Adding a mechanism for obtaining harvested metadata records unparsed, as Strings

landreev opened the pull request:
- Closes #284 (maybe?)

As the title says, this is a "model" implementation - primarily to illustrate what I want the final result to be on the application side. 
I expect that it will likely need to be implemented differently before the PR can be merged. I'm happy to work on a cleaner/better implementation.

landreev committed:
A simple implementation of a feature where an OAI client can be instructed to return harvested metadata records unparsed #284

landreev committed:
cosmetic #284

landreev committed:
cosmetic #284

pdurbin reviewed (approved):
I didn't run this but it seems reasonable to me.

poikilotherm reviewed (changes requested):
We should talk about this some more. I think there is a cleaner way to do this, maybe using the existing `CopyElement` or adding a new `RawElement`.

See https://github.com/gdcc/xoai/blob/83f9c1be130441f7be8817568bdfe3d1705579fa/xoai-common/src/main/java/io/gdcc/xoai/model/oaipmh/results/record/Metadata.java#L55-L55 and its usages in the data provider code parts. It seems the use cases are quite similar.

Also, this PR does not yet contain any unit tests. It might be interesting to see some benchmark tests as well to compare different approaches.

landreev commented:
Yes, sure, happy to talk about it this week. 
> I think there is a cleaner way to do this, maybe using the existing CopyElement or ...

The way I implemented it, using the existing `EchoElement` seemed like the easiest way to go about it at the moment. (What we did with `CopyElement` in the data provider was specifically in order to allow passing an InputStream to a metadata fragment cached in storage; and that seems like an overkill for what's needed here). But let me take another look. 

Yes, I will add a test shortly.

landreev committed:
renamed the new method for clarity/consistency #284

landreev committed:
A simple test for the new raw/unparsed metadata feature #284

landreev committed:
2 more checks added to the test #284

landreev committed:
cosmetic (mvn spotless:apply on the test class) #284

landreev commented:
I really like the `xoai-data-provider-tck` subsystem. You suggested at some point that this approach could be used for testing OAI functionality in Dataverse and I am considering implementing something similar. 
I am however getting one strange validation failure when running it locally: 
```
### Checking GetRecord response
...
FAIL:    Expected setSpec was missing from the response. The GetRecord response for identifier IdFgUHOANT did not contain a set specification for test1
```
(everything else appears to be passing). 
It is passing when the PR is tested here on github: 
```
PASS:    Expected setSpec was returned in the response
```
I'm puzzled as to what's going on - ?

landreev commented:
If you feel strongly about it, please let me know and I'll be happy to change the code to use `CopyElement` instead. Having taken another look at the classes, it doesn't seem like it's going to be much simpler/require fewer changes, but either way is ok with me.

landreev commented:
Can we talk about this sometime soon?

poikilotherm commented:
Let's do a VC call tomorrow! Ping me on Zulip, I don't have any appointments tomorrow. (I wasn't paying attention to this thread, halfway expecting you to ping me via DM again. Sorry!)

landreev committed:
Added the new StringElement for passing raw unparsed metadata #284

poikilotherm commented:
@landreev this is still a draft PR. Is that correct?

landreev commented:
> @landreev this is still a draft PR. Is that correct?

Sorry, this seriously slipped off of my radar. No, I thought I had checked in everything you suggested, the way we agreed back in March. I thought it was ready - but I obviously should have checked on it.

landreev commented:
@poikilotherm ... In other words, I would very much love to have this merged and made part of the release (I saw the [comment](https://github.com/gdcc/xoai/pull/283#issuecomment-3047992859) in #283). Please let me know if there is anything I can do to help make this happen.

poikilotherm commented on xoai-common/src/main/java/io/gdcc/xoai/model/oaipmh/results/record/Metadata.java:
```diff
@@ -78,6 +83,11 @@ public XOAIMetadata getXoaiMetadata() {
         else return null;
     }
 
+    public String asUnparsedString() {
+        if (element instanceof StringElement) return ((StringElement) element).asUnparsedString();
+        else return null;
+    }
+
```
Question: this is a very different API design compared to what's already here. Would it make sense to have a `write(StringWriter)` function instead? That way the control and resource allocation is in the hands of the caller. Also, we could define this as part of an interface and make the elements have this as well. So the actual "storage" classes can decide on the behavior when writing to an intended string target.

poikilotherm replied:
We decided via Zoom to not do this.

poikilotherm committed:
doc(common): improve asUnparsedString method documentation #284

- Add detailed Javadoc to clarify method behavior and return conditions for better maintainability and readability.

poikilotherm committed:
Merge branch 'branch-5.0' into fork/gdcc/284-unparsed-metadata-records

poikilotherm committed:
doc: added README entry for new unparsed metadata feature #284
